provider/cluster/manager.go

```diff
@@ -108,47 +112,73 @@ func (dm *deploymentManager) teardown() error {
 	}
 }
 
+func (dm *deploymentManager) force() error {
+	ch := make(chan error, 1)
+	select {
+	case dm.forcech <- ch:
+
+	case <-dm.lc.ShuttingDown():
+		return ErrNotRunning
+	}
+
+	select {
+	case err := <- ch:
+		return err
+	case <-dm.lc.ShuttingDown():
+		return ErrNotRunning
+	}
+}
+
+func (dm *deploymentManager) handleUpdate() <- chan error{
+	switch dm.state {
+	case dsDeployActive:
+		dm.state = dsDeployPending
+	case dsDeployComplete:
+		// start update
+		return dm.startDeploy()
+	case dsDeployPending, dsTeardownActive, dsTeardownPending, dsTeardownComplete:
+		// do nothing
+	}
+
+	return nil
+}
+
 func (dm *deploymentManager) run() {
 	defer dm.lc.ShutdownCompleted()
+	var shutdownErr error
 
-	allHostnames := util.AllHostnamesOfManifestGroup(*dm.mgroup)
-	reserveHostnamesCh := dm.hostnameService.ReserveHostnames(allHostnames, dm.lease.DeploymentID())
+	runch := dm.startDeploy()
 
-	var runch <-chan error
+	// TODO - somehow at this point if this deployment holds a hostname
+	// which is also requested for use by any other deployment with the same owner
+	// then the hostname should get moved back to the other deployment at this time
+	defer dm.hostnameService.ReleaseHostnames(dm.lease.DeploymentID())
 
-	var shutdownErr error
 loop:
 	for {
 		select {
-		case err := <-reserveHostnamesCh:
-			reserveHostnamesCh = nil
-			if err != nil {
-				deploymentCounter.WithLabelValues("reserve-hostnames", "err").Inc()
-				dm.log.Error("deploy hostname error", "state", dm.state, "err", err)
-				break loop
-			}
-			deploymentCounter.WithLabelValues("reserve-hostnames", "success").Inc()
-			defer dm.hostnameService.ReleaseHostnames(allHostnames)
-			runch = dm.startDeploy()
 
 		case shutdownErr = <-dm.lc.ShutdownRequest():
 			break loop
 
+		case forceResponse := <-dm.forcech:
+			if dm.state == dsDeployComplete {
+				newch := dm.handleUpdate()
+				if newch != nil {
+					runch = newch
```

Comment:
Put some documentation in here about how `handleUpdate()` will only return non-`nil` in cases that `runch` is already `nil`.

Also,

```go
if newch := dm.handleUpdate(); newch != nil {
  runch = newch
}
```